core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,651 @@
+        for (peer, &stake) in staked_nodes.iter() {
+            let Some(ag_addr) = self
+                .cluster_info
+                .lookup_contact_info(peer, |ci| ci.alpenglow())
+                .flatten()
+            else {
+                continue;
+            };
+            state.peers.insert(
+                *peer,
+                PeerData {
+                    stake: stake,
+                    address: ag_addr,
+                    relative_toa: [None; 4],
+                },
+            );
+            state.total_staked += stake;
+        }
+        trace!(
+            "Prepared for slot {slot}, total stake is {}",
+            state.total_staked
+        );
+    }
+
+    /// Collects votes and changes states
+    fn listener_thread(
+        self_state: Arc<[Mutex<SharedState>; 3]>,
+        should_exit: Arc<AtomicBool>,
+        cluster_info: Arc<ClusterInfo>,
+        mut epoch_specs: EpochSpecs,
+        socket: Arc<UdpSocket>,
+        command_sender: Sender<Command>,
+    ) {
+        socket
+            .set_read_timeout(Some(Duration::from_secs(1)))
+            .unwrap();
+        let id = cluster_info.id();
+        // Set aside enough space to fetch multiple packets from the kernel per syscall
+        let mut packets: Vec<Packet> = vec![Packet::default(); 1024];
+        loop {
+            // must wipe all Meta records to reuse the buffer
+            for p in packets.iter_mut() {
+                *p.meta_mut() = Meta::default();
+            }
+            // recv_mmsg should timeout in 1 second
+            let n = match recv_mmsg(&socket, &mut packets) {
+                Ok(n) => n,
+                Err(e) => {
+                    match e.kind() {
+                        std::io::ErrorKind::TimedOut | std::io::ErrorKind::WouldBlock => {
+                            0 // no packets received
+                        }
+                        _ => {
+                            error!("Got error {:?} in mock alpenglow RX socket operation, exiting thread", e.raw_os_error());
+                            return;
+                        }
+                    }
+                }
+            };
+            if should_exit.load(Ordering::Relaxed) {
+                return;
+            }
+            // we may have received no packets, in this case we can safely skip the rest
+            if n == 0 {
+                continue;
+            }
+
+            for pkt in packets.iter().take(n) {
+                if pkt.meta().size < MOCK_VOTE_HEADER_SIZE {
+                    continue;
+                }
+                let Some(pkt_buf) = pkt.data(..) else {
+                    continue;
+                };
+                let vote_pkt = MockVotePacketHeader::from_bytes(pkt_buf);
+                let pk = Pubkey::new_from_array(vote_pkt.sender);
+
+                let mut state = get_state_for_slot(self_state.as_slice(), vote_pkt.slot_number)
+                    .lock()
+                    .unwrap();
```

Comment:
grabbing this lock with every packet seems excessive. I wonder if we could cache all this information and refresh it before entering the loop or something

Reply:
well yes, this is somewhat inefficient. But holding this lock makes code a whole lot simpler. Given that we are not handling millions of packets here, we should be fine I believe. I'll make a bench-test to confirm this.